rclcpp/include/rclcpp/executor.hpp

```diff
@@ -76,9 +77,97 @@ class Executor
+  virtual void
+  add_callback_group(
+    rclcpp::CallbackGroup::SharedPtr group_ptr,
+    rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr,
+    bool notify = true);
+
+  /// Add vector of callback groups to an executor.
+  /**
+   * An executor can have zero or more callback groups which provide work during `spin` functions.
+   * You can pass a vector of shared pointers that point to callback groups and a shared pointer
+   * to the node base interface that owns the callback groups. Each callback group in the vector
+   * will be passed individually to the `add_callback_group(group_ptr, node_ptr, notify)`.
+   * This is a helper function that allows a user to pass a vector of callback groups and one
+   * node base interface in one call.
+   *
+   * \param[in] group_ptrs a vector of shared ptrs that point to callback groups
+   * \param[in] node_ptr a shared pointer that points to a node base interface
+   * \param[in] notify True to trigger the interrupt guard condition during this function. If
+   * the executor is blocked at the rmw layer while waiting for work and it is notified that a new
+   * callback group was added, it will wake up.
+   */
+  RCLCPP_PUBLIC
+  virtual void
+  add_callback_groups(
+    const std::vector<rclcpp::CallbackGroup::SharedPtr> & group_ptrs,
+    rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_ptr,
+    bool notify = true);
+
+  /// Get callback groups that belong to executor.
+  /**
+   * This function returns a vector of weak pointers that point to callback groups that were
+   * associated with the executor.
+   * The callback groups associated with this executor may have been added with
+   * `add_callback_groups`, or added when a node was added to the executor with `add_node`, or
+   * automatically added when it created by a node already associated with this executor and the
+   * automatically_add_to_executor_with_node parameter was true.
+   *
+   * \return a vector of weak pointers that point to callback groups that are associated with
+   * the executor
+   */
+  RCLCPP_PUBLIC
+  virtual std::vector<rclcpp::CallbackGroup::WeakPtr>
+  get_callback_groups();
+
+  /// Remove a callback group from the executor.
+  /**
+   * The callback group is removed from the executor.
+   * If the callback group removed was the last callback group from the node
+   * that is associated with the executor, the interrupt guard condition
+   * is triggered and node's guard condition is removed from the executor
+   *
+   * \param[in] group_ptr Shared pointer to the callback group to be added.
+   * \param[in] notify True to trigger the interrupt guard condition during this function. If
+   * the executor is blocked at the rmw layer while waiting for work and it is notified that a
+   * callback group was removed, it will wake up.
+   * \throw std::runtime_error if the callback group is not associated with the executor
+   */
+  RCLCPP_PUBLIC
+  virtual void
+  remove_callback_group(
```

Comment:
@peterpena and I talked about this offline and I think we should change this so that you may only ever remove a callback group with this method which was explicitly added with `add_callback_group`. It should not be possible to remove callback groups added with `add_node` and it should not be possible to implicitly remove a node that was added with `add_node` by removing its only callback group. These side-effects and interactions make these functions more complicated than they need to be.

This will also affect `remove_node` and `has_node`.